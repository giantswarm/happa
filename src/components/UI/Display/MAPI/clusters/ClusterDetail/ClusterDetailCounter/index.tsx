import { Box, Text } from 'grommet';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import NotAvailable from 'UI/Display/NotAvailable';

import ClusterDetailWidgetOptionalValue from '../ClusterDetailWidgetOptionalValue';

function pluralizeLabel(count: number | undefined, base: string) {
  if (count === 1) {
    return base;
  }

  return `${base}s`;
}

const Label = styled(Text)`
  text-transform: uppercase;
`;

interface IClusterDetailCounterProps
  extends React.ComponentPropsWithoutRef<typeof Box> {
  label: string;
  value?: number;
  pluralize?: boolean;
}

const ClusterDetailCounter: React.FC<IClusterDetailCounterProps> = ({
  label,
  value,
  pluralize,
  ...props
}) => {
  const formattedLabel = pluralize ? pluralizeLabel(value, label) : label;

  let a11yLabel = `Loading ${formattedLabel}...`;
  if (typeof value === 'number' && value >= 0) {
    a11yLabel = `${value} ${formattedLabel}`;
  } else if (typeof value === 'number' && value < 0) {
    a11yLabel = `${formattedLabel} not available`;
  }

  return (
    <Box align='center' basis='100px' flex={{ grow: 0, shrink: 1 }} {...props}>
      <Box margin={{ bottom: 'xsmall' }}>
        <ClusterDetailWidgetOptionalValue
          value={value}
          replaceEmptyValue={false}
          loaderHeight={22}
          loaderWidth={30}
        >
          {(counterValue) => (
            <Text size='xxlarge' aria-label={a11yLabel}>
              {counterValue === -1 ? <NotAvailable /> : counterValue}
            </Text>
          )}
        </ClusterDetailWidgetOptionalValue>
      </Box>
      <Label color='text-weak' size='small'>
        {formattedLabel}
      </Label>
    </Box>
  );
};

ClusterDetailCounter.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.number,
  pluralize: PropTypes.bool,
};

export default ClusterDetailCounter;
