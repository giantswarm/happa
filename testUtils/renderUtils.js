import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
