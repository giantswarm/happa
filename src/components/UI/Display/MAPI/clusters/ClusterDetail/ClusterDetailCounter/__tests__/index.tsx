import { renderWithTheme } from 'test/renderUtils';

import ClusterDetailCounter from '..';

describe('ClusterDetailCounter', () => {
  it('renders a simple counter', () => {
    const { container } = renderWithTheme(ClusterDetailCounter, {
      label: 'dog',
      value: 1,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a counter with a pluralized label', () => {
    const { container } = renderWithTheme(ClusterDetailCounter, {
      label: 'dog',
      value: 35,
      pluralize: true,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
