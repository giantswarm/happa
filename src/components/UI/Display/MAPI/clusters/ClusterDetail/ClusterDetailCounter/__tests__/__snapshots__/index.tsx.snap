// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterDetailCounter renders a counter with a pluralized label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hSiATR"
>
  <div
    class="sc-gKckTs izRacH"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 eUwwzg"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eITwkI"
      >
        <div
          class="sc-ctqRej pDxxo sc-cdJjZP gDgffG"
        >
          <span
            aria-label="35 dogs"
            class="StyledText-sc-1sadyjn-0 kQovFR"
          >
            35
          </span>
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 bTNMGz sc-hFLmgA bLwsQw"
      >
        dogs
      </span>
    </div>
  </div>
</div>
`;

exports[`ClusterDetailCounter renders a simple counter 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hSiATR"
>
  <div
    class="sc-gKckTs izRacH"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 eUwwzg"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eITwkI"
      >
        <div
<<<<<<< HEAD
          class="sc-dFtzdE eKjOri sc-cdJjZP gDgffG"
=======
          class="sc-ctqRej pDxxo sc-cdJjZP gDgffG"
>>>>>>> 0ef72451 (Update snapshots)
        >
          <span
            aria-label="1 dog"
            class="StyledText-sc-1sadyjn-0 kQovFR"
          >
            1
          </span>
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 bTNMGz sc-hFLmgA bLwsQw"
      >
        dog
      </span>
    </div>
  </div>
</div>
`;
